### PR TITLE
stream_flash minor fixes

### PR DESCRIPTION
**Pull request by de-nordic**
The batch of changes that include:
 1) Use flash_get_parameters to get erase value for stream_flash_buffered_write, where read of erased flash has been used previously;
 2) The ctx->buf_bytes check, within flash_sync, has been moved outside CONFIG_STREAM_FLASH_ERASE conditional code, as it is universal check whether there is any work to do;
 3) The update, within flash_sync,  of flash_stream_ctx bytes_written and buf_bytes is not done when verification callback fails;
 4) The stream_flash_erase_page will only update last_erased_page_start_offset if successful.

### CODE REVIEW

File: subsys/storage/stream/stream_flash.c [hunk 1]
```diff
@@ -42,6 +41,8 @@ int stream_flash_erase_page(struct stream_flash_ctx *ctx, off_t off)
 
 	if (rc != 0) {
 		LOG_ERR("Error %d while erasing page", rc);
+	} else {
+		ctx->last_erased_page_start_offset = page.start_offset;
```
Comment (hakonfam) [inline subsys/storage/stream/stream_flash.c:45]: Could we add a unit test case for this?

File: subsys/storage/stream/stream_flash.c [hunk 2]
```diff
@@ -97,6 +99,7 @@ static int flash_sync(struct stream_flash_ctx *ctx)
 		rc = ctx->callback(ctx->buf, ctx->buf_bytes, write_addr);
 		if (rc != 0) {
 			LOG_ERR("callback failed: %d", rc);
+			return rc;
```
Comment (hakonfam) [inline subsys/storage/stream/stream_flash.c:102]: Could we add a unit test case for this?

**Review by hakonfam — APPROVED** [review]
LGTM, just wondering if we should have test cases for the fixes to avoid regression.

File: subsys/storage/stream/stream_flash.c (hunk 1 above)
Comment (de-nordic) [reply subsys/storage/stream/stream_flash.c:45]: Yes, we should. Sorry I have completely forgotten to write tests for this while stuck on on other tasks. I will add tests for the changes.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]
fix the last commit message title.

**Comment by de-nordic** [comment]
@hakonfam I am still working on the tests but in meantime I have added this to the stack: https://github.com/zephyrproject-rtos/zephyr/pull/32488/commits/ba6782601fc6ccde968b07add88b235f4702f3b6, can you look at that one?

**Comment by hakonfam** [comment]
> @hakonfam I am still working on the tests but in meantime I have added this to the stack: [ba67826](https://github.com/zephyrproject-rtos/zephyr/commit/ba6782601fc6ccde968b07add88b235f4702f3b6), can you look at that one?

Good catch! Again a unit test verifying the fix would be great. What do you think about providing `fill_length` as an argument to `flash_sync`? Maybe even the whole handling of the filling of the tail? As we already see, the current split of this makes it prone for bugs where the state is not updated correctly.

**Comment by de-nordic** [comment]
> Again a unit test verifying the fix would be great. 

I am working on them. Unfortunately this requires overriding flash API to make functions fail at specific moments.

> What do you think about providing `fill_length` as an argument to `flash_sync`? Maybe even the whole handling of the filling of the tail? As we already see, the current split of this makes it prone for bugs where the state is not updated correctly.

The buffer alignments should happen closer to the `flash_write`, so yes probably moving them to `flash_sync` would be good idea.
That would also reduce arithmetic on correcting the ctx after failed operations as that can also be done closer to the source of failure.

**Comment by hakonfam** [comment]
> Would you be willing to do such change?

Yes, I will get this done.

**Comment by de-nordic** [comment]
> 
> 
> > Would you be willing to do such change?
> 
> Yes, I will get this done.

I think we should have some rfc for the stream_flash changes.
While fixing the issues I have found that it is hard to recover from failed write attempts.
For example: when you attempt to write N lengths of buffer and buffer already has some data, in case of failure you have no way to actually find out at which byte, at which buffer the write failed; this means that when write of first buffer fails, you have probably lost data that you have no control over, because it has been placed in buffer with some previous calls to write.

**Review by nvlsianpu — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@de-nordic Thanks for `a bad weather` testcases.

**Pull request merged**